### PR TITLE
Add a min-max heap implementation that can be used to back a priority queue

### PR DESCRIPTION
**Pull request by AquaGeek**
### Description

This is an extraction of the heap implementation hammered out in #51.

### Detailed Design

One of the big advantages of using a min-max heap is that we don't need to bifurcate into a min heap or max heap and keep track of which kind is used — you can pop min or max, as needed. This makes it very useful as the backing storage for a double-ended priority queue.

```swift
public struct Heap<Element: Comparable> {
    public var isEmpty: Bool
    public var count: Int

    // Read-only view into the underlying array
    public var unordered: [Element]

    // Initializers
    public init()
    public init<S: Sequence>(_ elements: S)

    // Mutations
    public mutating func insert(_ element: Element)
    public mutating func insert<S: Sequence>(contentsOf: S)
    public func min() -> Element?
    public func max() -> Element?
    public mutating func popMin() -> Element?
    public mutating func popMax() -> Element?
    public mutating func removeMin() -> Element
    public mutating func removeMax() -> Element
}

// Array literal
extension Heap: ExpressibleByArrayLiteral {
    public init(arrayLiteral elements: Element...)
}

// Iteration
extension Heap {
    public struct Iterator: Sequence, IteratorProtocol {}

    public var ascending: Iterator
    public var descending: Iterator
}
```

### Future Directions

* Add `replaceMin` and `replaceMax` (@timvermeulen has [a branch](https://github.com/timvermeulen/swift-collections/tree/priority-queue-replace) of this already)
* Implement a priority queue on top of this (that work will be picked back up in #51)

### Documentation

The public APIs have largely been documented. An overview document has been added to the Documentation directory that is adapted from the [API review](https://forums.swift.org/t/api-review-priorityqueue-apis/49932) and code comments.

### Testing

There are unit tests for the added `Heap` type.

### Performance

Performance tests have been added (including comparable tests against a `CFBinaryHeap`). We may want to revisit the library JSON to ensure we have the desired comparisons defined. (Can somebody with a better understanding of how all of that works/is structured take a look? Maybe @lorentey?)

### Source Impact

This is purely additive. No existing APIs were changed, deprecated, or removed.

### Checklist

- [X] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-collections)
- [X] My contributions are licensed under the [Swift license](https://swift.org/LICENSE.txt).
- [X] I've followed the coding style of the rest of the project.
- [X] I've added tests covering all new code paths my change adds to the project (to the extent possible).
- [X] I've added benchmarks covering new functionality (if appropriate).
- [X] I've verified that my change does not break any existing tests or introduce unexpected benchmark regressions.
- [x] I've updated the documentation (if appropriate).

### CODE REVIEW

**Comment by AquaGeek** [comment]
OK, [`Heap.md`](https://github.com/apple/swift-collections/pull/61/files#diff-cffffb935bfba8cbdaf7e223ba42b053ae26bfc6457f5635bd94b609f28f44fa) has been added to the `Documentation` directory. This should be ready for a full review.

File: Documentation/Heap.md
Comment (hassila) [inline Documentation/Heap.md:119]: Was there supposed to be any performance data under this header?

Comment (AquaGeek) [reply Documentation/Heap.md:119]: Ooops. I added that because it was in most of the other docs. I didn't end up filling it in, as it's covered elsewhere in here. Thoughts?

Comment (hassila) [reply Documentation/Heap.md:119]: Perhaps just move the time complexity characteristics from the overview section there and basically list the operations with their time complexity characteristics? 

There is no precedent for adding performance graphs, but in general I think that could be useful if combined with hardware information - when looking at data structure repos, it is one of the first obvious questions for me at least, just to get an indication of approximate level of performance (as using terms as "fast" have very different meanings to different people in my experience - and algorithmic complexity doesn't say anything about constant overhead which might be troublesome depending on use case). So I'd consider attaching the perf data for just this data structure.

File: Benchmarks/CppBenchmarks/src/BinaryHeap.m
Comment (kylemacomber) [inline Benchmarks/CppBenchmarks/src/BinaryHeap.m:65]: To make the benchmark a fairer comparison, let's insert the `NSInteger`s directly into the `CFBinaryHeap` (rather than `NSNumber`s). All we'd need to do is update the retain/release hooks to be no-ops and update insert/remove/compare to do the appropriate pointer casting.

As it stands the `objc_msgSend` overhead is probably convoluting the benchmark.

File: Sources/PriorityQueueModule/Heap+Iterator.swift
Comment (kylemacomber) [inline Sources/PriorityQueueModule/Heap+Iterator.swift:28]: I don't _think_ there's any reason to make this initializer (or the `IterationDirection` enum) public. (It can be marked as `@inlinable` even if it's internal).

File: Benchmarks/CppBenchmarks/src/BinaryHeap.m
Comment (AquaGeek) [reply Benchmarks/CppBenchmarks/src/BinaryHeap.m:65]: Wouldn't we need to manually manage the memory for the integers, as the address of `value` is only valid for the scope of the method?

File: Documentation/Heap.md
Comment (Dante-Broggi) [inline Documentation/Heap.md:103]:
```suggestion
We also have non-optional flavors that assume the heap isn't empty, `removeMin()` and `removeMax()`.
```

File: Benchmarks/CppBenchmarks/src/BinaryHeap.m
Comment (AquaGeek) [reply Benchmarks/CppBenchmarks/src/BinaryHeap.m:65]: (ping, @kylemacomber)

Comment (kylemacomber) [reply Benchmarks/CppBenchmarks/src/BinaryHeap.m:65]: A primitive type like `NSInteger` is passed by value—there's no heap allocation that needs to be managed. CF collection types support elements that are the size of a pointer. `NSInteger` is the exact same size as a pointer, so we can insert it into the CFBinaryHeap and just implement retain/release to be no-ops (since, again, there's no heap allocation that needs to be reference counted).

Comment (AquaGeek) [reply Benchmarks/CppBenchmarks/src/BinaryHeap.m:65]: Ah! Got it — cast the integer to a pointer when inserting and back again when getting the minimum value/doing comparisons. I had assumed we would need to insert a pointer to the integer into the heap (hence my question above), so I started `malloc`/`free`ing.

**Comment by AquaGeek** [comment]
I've updated the `CFBinaryHeap` benchmark. Using raw integers instead of `NSNumber`s significantly altered the performance comparison. Beyond 256 elements, `CFBinaryHeap` is faster than `Heap` for `popMin`.

CFBinaryHeap Updated Comparison

File: Documentation/Heap.md
Comment (AquaGeek) [reply Documentation/Heap.md:119]: I added a table with the performance of the various operations. I'll defer to @kylemacomber and @lorentey on the point about performance graphs. I agree that they can be helpful, but I also wonder what the baseline of comparison one would want to see in them — Comparing to `Foundation`/`Core Foundation` types? Comparing to C++'s stdlib types? Just average per-element operation time (i.e. only show the graphs for `Heap` itself)?

Comment (hassila) [reply Documentation/Heap.md:119]: Thanks! Speaking from my experience I’d be happy with just the graphs for Heap itself. It gives some orientation and if I really care about performance at that level, I’d already have some notion of how other similar structures perform.  I think the most important thing is to give the magnitude of performance numbers to give context - as we all have very widely different views of “fast” - this anchors that notion.

**Comment by hassila** [comment]
> I've updated the `CFBinaryHeap` benchmark. Using raw integers instead of `NSNumber`s significantly altered the performance comparison. Beyond ~256 elements, `CFBinaryHeap` is faster than `Heap` for both `popMin` and `popMax`.
> 
> CFBinaryHeap Updated Comparison

Thanks a lot, interesting!

I’d just want to (very shortly) make the point here that it’s always faster, CF only has a higher constant creation time (I know you know, but it was related to the point of how to display performance numbers in the related PR) - this conflation of numbers leads to unprecise characterizations of perf at times. Anyway, I’ll be silent about that now :-).

File: Documentation/Heap.md
Comment (AquaGeek) [reply Documentation/Heap.md:119]: Added

**Comment by AquaGeek** [comment]
@kylemacomber How do you want to handle this? The code was already mostly reviewed in the other PR; should @timvermeulen take another look?

Also, do we want to merge this into `main` or into a `priority-queue` branch?

**Comment by kylemacomber** [comment]
@lorentey is the final boss 😉

**Comment by AquaGeek** [comment]
@lorentey Any update on when you might be able to look at this? 🙏

**Comment by lorentey** [comment]
I'm on it! It's slow going, but I'll submit the review soon! (Things are looking good.)

**Comment by lorentey** [comment]
Quick update on our plans: I'd like to release version 1.0 of this package relatively soon with the existing three types (to signal that we're committing to source stability), then quickly follow it up with version 1.1 that includes this work as well as anything else that's ready.

However, we don't need to wait on the 1.0 tag to merge new work to `main` -- we can just tag 1.0 from a release branch, like we do for Swift releases. So we can land this as soon as I get my act together and finish the review. :-)

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [inline Sources/PriorityQueueModule/Heap.swift:519]: Nit: put spaces around all binary operators -- `..<` and `...` shouldn't be special.

```suggestion
    for idx in (0 ..< (storage.count / 2)).reversed() {
```

File: Tests/PriorityQueueTests/HeapTests.swift
Comment (lorentey) [inline Tests/PriorityQueueTests/HeapTests.swift:131]: FIXME: Combinatorial tests would work great on these!

This doesn't need to be addressed before this lands.

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [inline Sources/PriorityQueueModule/Heap.swift:14]: This is useful information to have in an "Implementation Details" section, but it doesn't tell the reader what a heap is and why they would want to use it. We should add an introduction that is useful for a wider audience.

Something like this would be a good start:

> A collection of comparable items arranged in a way that allows efficient access to its minimal and maximal elements.
>
> Heaps can be used to efficiently provide sorted access to items that arrive incrementally and out of order. For example, a server application may use a heap to store its queue of incoming client transactions, processing them in order of decreasing priority.
>
> Heaps provide highly optimized implementations for the following operations:
>
> - `insert(_:)`: Inserts a new element in O(log(`count`)) time.
> - `min()`/`max()`: Retrieve (but does not remove) the minimum/maximum item in O(1) time.
> - `removeMin()`/`removeMax()`: Remove and return the minimum/maximum item in O(log(`count`)) time.
>
> \<Some toy sample code illustrating these operations\>

**Review by lorentey — APPROVED** [review]
I think this is ready to land, with a small handful of nits that would be useful to resolve before merging.

I also noted several larger FIXMEs/experiments that we should eventually get around to implementing, but they need not be done before merging this PR.

File: Sources/PriorityQueueModule/Heap.swift
Comment (AquaGeek) [reply Sources/PriorityQueueModule/Heap.swift:421]: It was determined in the original PR that doing it this way was [more performant](https://github.com/apple/swift-collections/pull/51#issuecomment-860469370).

File: Sources/PriorityQueueModule/Heap+Iterator.swift
Comment (AquaGeek) [reply Sources/PriorityQueueModule/Heap+Iterator.swift:14]: Yeah, I think someone brought that up in the other PR as a performance win (i.e. avoid the branch). @CTMacUser did this in [his PR](https://github.com/AquaGeek/swift-collections/pull/3/files#diff-b04be4dc0d13a0aa819f534acfeb2c81f2368c46ff4d814a24313d8e3ce9fc7c) for the priority queue; maybe I'll do the same here for the heap.

File: Sources/PriorityQueueModule/Heap+Iterator.swift
Comment (AquaGeek) [inline Sources/PriorityQueueModule/Heap+Iterator.swift:47]:
```suggestion
    return Iterator(_base: self, direction: .ascending)
```

File: Sources/PriorityQueueModule/Heap+Iterator.swift
Comment (AquaGeek) [inline Sources/PriorityQueueModule/Heap+Iterator.swift:53]:
```suggestion
    return Iterator(_base: self, direction: .descending)
```

File: Benchmarks/Benchmarks/PriorityQueueBenchmarks.swift
Comment (AquaGeek) [reply Benchmarks/Benchmarks/PriorityQueueBenchmarks.swift]: There was a _lot_ of discussion about these tests in the original PR. I accepted your suggestion, limiting the measurement to only the pop operation.

File: Benchmarks/CppBenchmarks/include/BinaryHeap.h
Comment (AquaGeek) [reply Benchmarks/CppBenchmarks/include/BinaryHeap.h:22]: Yeah. I started writing a C++ benchmark for comparison but haven't been able to finish that yet. I'd be happy for any assistance improving the comparative benchmarks here from anyone; I'm quickly approaching the limits of my understanding of perf.

**Comment by AquaGeek** [comment]
OK, updated things. I think the only things left from the "need to fix before merging this" bucket are the documentation changes.

File: Benchmarks/Benchmarks/PriorityQueueBenchmarks.swift
Comment (lorentey) [reply Benchmarks/Benchmarks/PriorityQueueBenchmarks.swift]: Ah, sorry I missed those. Generally I think it's a good idea for these microbenchmarks to each concentrate on a single operation -- it makes it easier to compare alternative approaches and to interpret results.

File: Benchmarks/CppBenchmarks/include/BinaryHeap.h
Comment (lorentey) [reply Benchmarks/CppBenchmarks/include/BinaryHeap.h:22]: I can have a go at it relatively soon.

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap.swift]: Wow, that's terrible. This is another FIXME for later -- there is no reason `Array.swapAt` needs to be any slower than doing this.

File: Sources/PriorityQueueModule/CMakeLists.txt
Comment (lorentey) [inline Sources/PriorityQueueModule/CMakeLists.txt:14]:
```suggestion
  Heap+OrderedViews.swift)
```

**Comment by lorentey** [comment]
The doc updates we can also do as followup work. Let's land this! 🚀 

Do you need me to help resolve the conflicts?

**Comment by AquaGeek** [comment]
> The doc updates we can also do as followup work. Let's land this! 🚀

Sounds good!

> Do you need me to help resolve the conflicts?

I can do it. Do you want me to rebase on `main` or pull it in to resolve the conflicts?

**Comment by AquaGeek** [comment]
Merge conflicts have been resolved. We should be good to go! 🚀

Do you want me to create GitHub issues for the follow-up changes?

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
> Do you want me to create GitHub issues for the follow-up changes?

Oh, that's a good idea -- let me do that though; it seems unfair to ask you to file issues for my loosely explained FIXMEs. 🙈

**Comment by lorentey** [comment]
```
19:05:20 /home/buildnode/jenkins/workspace/pr-swift-collections-linux/branch-main/swift-collections/Benchmarks/CppBenchmarks/src/BinaryHeap.m:12:9: fatal error: 'Foundation/Foundation.h' file not found
19:05:20 #import <Foundation/Foundation.h>
19:05:20         ^~~~~~~~~~~~~~~~~~~~~~~~~
19:05:20 1 error generated.
```

Ah, this makes sense -- I think we'll need to add `#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)` around these benchmarks.

(Which is super annoying, as it makes the benchmark library json platform dependent. Argh -- this is another FIXME for later.)

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap.swift:14]: https://github.com/apple/swift-collections/issues/74

File: Tests/PriorityQueueTests/HeapTests.swift
Comment (lorentey) [reply Tests/PriorityQueueTests/HeapTests.swift:131]: Issue filed as https://github.com/apple/swift-collections/issues/73

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap.swift]: Issue filed as https://github.com/apple/swift-collections/issues/72 (in addition to the existing Swift issue https://bugs.swift.org/browse/SR-14778 -- thanks very much for filing that!)

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap.swift:410]: Issue filed as https://github.com/apple/swift-collections/issues/71

File: Sources/PriorityQueueModule/Heap.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap.swift:154]: Issue filed as https://github.com/apple/swift-collections/issues/75

File: Sources/PriorityQueueModule/Heap+Iterator.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap+Iterator.swift]: Issue filed as https://github.com/apple/swift-collections/issues/70

File: Sources/PriorityQueueModule/Heap+Invariants.swift
Comment (lorentey) [reply Sources/PriorityQueueModule/Heap+Invariants.swift]: Issue filed as https://github.com/apple/swift-collections/issues/69

**Comment by lorentey** [comment]
OK, I filed issues for all my FIXMEs:

https://github.com/apple/swift-collections/issues?q=is%3Aissue+is%3Aopen+label%3AHeap

As usual with these, some of them are suggestions for optimizations that may turn out to be unnecessary or even harmful. The benchmarks will be the judge. 😝

**Comment by lorentey** [comment]
@swift-ci test

**Pull request merged**
